Large gas reservoirs and filled pipe mapping

IProduceWidgets opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

requires https://github.com/space-wizards/RobustToolbox/pull/5414

## About the PR
<!-- What did you change in this PR? -->
Also adds large static gas reservoirs which can be de/constructed, but not unanchored and moved.
![image](https://github.com/user-attachments/assets/d68662b6-cbeb-4f9f-9d50-ca5c11973113)

Probably most impactfully, this PR adds save / load support for PipeNets via the PipeFill component & system.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->

Having immovable gas reservoirs allows us to dramatically increase the stored gas on the station. By being immovable they are much harder to grief with than other sources of gas storage.

Similarly, the ability to map pre-filled PipeNets dramatically increases the roundstart gas available to the station for distro (and in a way that scales very seemlessly with station size!), and combined these two changes may allow us to move away from gas miners.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.
-->

## Requirements
<!-- 
Due to influx of PR's we require to ensure that PR's are following the correct guidelines.

Please take a moment to read these if its your first time.

Check the boxes below to confirm that you have in fact seen these (put an X in the brackets, like [X]):
-->
- [x] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

Entities with NodeContainers which have a Pipe node should also have the `- type: PipeFill` component. It will be added automatically at runtime, but it may cause tests to fail if you have a prototype with a Pipe node that does not have it.

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up. Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl:
- add: huge gas reservoirs which cannot be moved.
- tweak: distro or other pipenets may now start with gas in them.

slarticodefast commented:
You might have to rebalance the grand lottery for this since the frezon canister should be worth around 60k spesos now.

ImRenQ commented:
Can't large gas canisters be not a replacement, but just an additional option?

Ilya246 commented:
some criticisms:
- this is WAY overtuned, they should probably *at most* ever contain 2500L
- the new prices for the canisters are also way overtuned; also remember that, out of the cost, 1000 spesos are actually for the canister

deltanedas commented:
10m^3 would make sense for a bluespace canister not a random plasteel one

EthanQix commented:
Maybe I'm minsunderstanding the mole/volume/pressure mechanics, but couldn't you just increase the default pressure in the canisters in order to fit more moles in the same volume ? So it would pressurize more tiles with one canister ?

IProduceWidgets commented:
> Maybe I'm minsunderstanding the mole/volume/pressure mechanics, but couldn't you just increase the default pressure in the air canisters in order to fit more moles in the same volume ? So it would pressurize more tiles with one canister ?

yes, but theres no easy way to pressurize anything over 4.5kkpa

EthanQix commented:
> > Maybe I'm minsunderstanding the mole/volume/pressure mechanics, but couldn't you just increase the default pressure in the air canisters in order to fit more moles in the same volume ? So it would pressurize more tiles with one canister ?
> 
> yes, but theres no easy way to pressurize anything over 4.5kkpa. Also they eventually break

yeah, my point here is why is there a hardcoded 4.5MPa maximum ? And why that specific, somewhat low number ? Is it an atmos engine limitation ?

deltanedas commented:
you can trivially make it 9mpa with a volpump, also there is no breakage mechanic you can have a 10gpa can and its fine

EthanQix commented:
> you can trivially make it 9mpa with a volpump, also there is no breakage mechanic you can have a 10gpa can and its fine

ah so the air canisters can be set to 10x as many moles with the same volume and there won't be a problem ? ^^

slarticodefast commented:
> ah so the air canisters can be set to 10x as many moles with the same volume and there won't be a problem ? ^^

That means the pressure in the canister would be 10x as high. With that you could easily overpressurize a room to the point where it damages players. Also with the pump limit being at 9kPa there would be no way to refill them. For the future there is however a softclogging change for pumps planned, giving them an efficiency curve instead of a hard maximum limit.

EthanQix commented:
True but now this PR should be as easy as just modifying the default mole value in air canisters instead of modifying every canister volume stat ^^

IProduceWidgets commented:
update: I added large gas resevoirs

Sarahon commented:
I don't mind this tbh. Air alarms are kinda slow as hell at times and there's a risk when you max out distro just to fix spacing. Allowing us to utilise canisters more would fix this obvious do or die scenario. Giving us a reason to create more reservoirs without having to dismantle and rebuild atmos and more. 

A situation you can run into is that spacing isn't fixed fast enough. I usually set distro itself to 500kpa so it still has a lot of gas in the pipes. I rarely set up a reservoir because there's really just not enough space for them (oasis reservoir is an example of how big they'd need to get, and even then it's not that useful in terrible spacing). But if you set the pipe pressure to max, you can space the whole of distro - or overpressure a room if you forget to turn off the fill option(while also leaking distro). It could be fun! Not sure.

Partmedia commented:
You may be able to find discussion about this before from this old PR: https://github.com/space-wizards/space-station-14/pull/11242

Why is the `GasPipeFill` component and system included in this PR? This feels like a different issue that is better split into a separate PR per our [pull request guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).

IProduceWidgets commented:
> You may be able to find discussion about this before from this old PR: #11242
> 
> Why is the `GasPipeFill` component and system included in this PR? This feels like a different issue that is better split into a separate PR per our [pull request guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).

Its soap, but yes, I appear to have buried the lead here on what has become PipeNet save/load support.

Partmedia commented:
> Its soap, but yes, I appear to have buried the lead here on what has become PipeNet save/load support.

What do you mean that "its soap?" These sound like separate changes to me, and it would be easier to discuss, merge, and back out if something goes awry if these are separate commits, per our pull request guidelines. Thanks in advance.

IProduceWidgets commented:
> > Its soap, but yes, I appear to have buried the lead here on what has become PipeNet save/load support.
> 
> What do you mean that "its soap?" These sound like separate changes to me, and it would be easier to discuss, merge, and back out if something goes awry if these are separate PRs per our pull request guidelines. Thanks in advance.

I had to support being able to fill pipes to fill the pipes.

Partmedia commented:
I would prefer if the gas canister change and the gas reservoir/filled pipe changes were split into separate PRs. This is better aligned with our pull request guidelines and will make it easier to get through design and code review. If you want to keep them together expect a longer review time. Thanks.

IProduceWidgets commented:
Split the canister changes to #31667 since it wasn't much trouble.

I also addressed the cargo lottery stuff on that PR by reducing the odds of the valuable cans. Im going to move all conversation about canisters, except as it pertains to mapping them as a ways to (somewhat poorly) fill pipenets to that PR.

Thanks.

IProduceWidgets commented:
> I was about to work on a large gas reservoir before finding this PR
> 
> Are you still working on this? Could you spin off the reservoirs so we can map them for shuttles and such?

The sprites of course you can use. Unfortunately the resevoirs are almost intrinsically tied to the filled pipe mapping because they are implemented as simple pipes because they don't make sense to be "portable" (as in used with connector ports). However, I would also direct you to #31667 which does something similar by increasing the volume of normal gas canisters, as well as adding new sprites for "smaller" gas canisters which have our current canister volumes.

IProduceWidgets commented:
Nah, I can go ahead a close this one. I'll reopen it if I get around to poking serialization or give up and use a marker ent.
